File: chapter10/src/main/java/com/example/chapter10/part1/RoundRectShapeView2.java
package com.example.chapter10.part1;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RoundRectShape;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import com.example.chapter10.Utils;

/**
 * RoundRectShape 实现中间带有镂空矩形的圆角矩形并且中间的镂空矩形也带有圆角
 *
 * @author wangzhichao
 * @date 2019/10/11
 */
public class RoundRectShapeView2 extends View {
    private ShapeDrawable drawable;

    public RoundRectShapeView2(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        // 外围矩形的各个角的圆角
        float[] outRadii = new float[]{
                Utils.dp2px(12), Utils.dp2px(12),
                Utils.dp2px(24), Utils.dp2px(24),
                0, 0,
                0, 0};
        // 内部矩形与外部矩形各边的间距
        RectF inset = new RectF(Utils.dp2px(10), Utils.dp2px(10), Utils.dp2px(10), Utils.dp2px(10));
        // 内部矩形的各个角的圆角
        float[] innerRadii = new float[]{
                0, 0,
                0, 0,
                Utils.dp2px(12), Utils.dp2px(12),
                Utils.dp2px(24), Utils.dp2px(24)
        };
        RoundRectShape roundRectShape = new RoundRectShape(outRadii, inset, innerRadii);
        drawable = new ShapeDrawable(roundRectShape);
        drawable.setBounds(new Rect(Utils.dp2px(50), Utils.dp2px(50), Utils.dp2px(200), Utils.dp2px(100)));
        drawable.getPaint().setColor(Color.YELLOW);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        drawable.draw(canvas);
    }
}

File: chapter10/src/main/java/com/example/chapter10/part1/RoundRectShapeView2.kt
package com.example.chapter10.part1

import android.content.Context
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.RoundRectShape
import android.graphics.*
import android.util.AttributeSet
import android.view.*
import com.example.common.dp

/**
 * RoundRectShape 实现中间带有镂空矩形的圆角矩形并且中间的镂空矩形也带有圆角
 *
 * @author wangzhichao
 * @date 2019/10/11
 */
class RoundRectShapeView2(context: Context, attrs: AttributeSet?) : View(context, attrs) {
    private val drawable: ShapeDrawable
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawable.draw(canvas)
    }

    init {
        // 外围矩形的各个角的圆角
        val outRadii = floatArrayOf(
            12.dp, 12.dp,
            24.dp, 24.dp, 0f, 0f, 0f, 0f)
        // 内部矩形与外部矩形各边的间距
        val inset = RectF(10.dp, 10.dp, 10.dp, 10.dp)
        // 内部矩形的各个角的圆角
        val innerRadii = floatArrayOf(0f, 0f, 0f, 0f,
            12.dp, 12.dp,
            24.dp, 24.dp,)
        val roundRectShape = RoundRectShape(outRadii, inset, innerRadii)
        drawable = ShapeDrawable(roundRectShape)
        drawable.bounds = Rect(50.dp.toInt(),
            50.dp.toInt(),
            200.dp.toInt(),
            100.dp.toInt())
        drawable.paint.color = Color.YELLOW
    }
}